Add support for UWP apps from base library

davor-skontra opened the pull request:
The author from the base repo added support for UWP applications, Those changes came with something of a refactor. (Some methods were moved around into a new file)

I went through it, found the methods you originally modified to stop the lib from panicking and applied those changes again.

Did some testing against my own work in progress project and it worked without issues.

Note I'm not an experienced Rust programmer so keep an eye for something stupid only a C# dev would do. :)

z3k-io committed:
Merge changes from original repo

davor-skontra committed:
Implement LGUG2Z panic removing

davor-skontra committed:
Unwrap calls in main

This is an example intended for the original repo so no need sprucing
this up. Would consider nuking the whoöe examples folder were it up to
me.

davor-skontra committed:
Remove examples

LGUG2Z commented on src/uwp_apps.rs:
```diff
@@ -0,0 +1,56 @@
+
+pub fn get_uwp_icon(process_path: &str) -> Result<RgbaImage, Box<dyn Error>> {
+    let icon_path = get_icon_file_path(process_path)?;
+    println!("Icon path: {}", icon_path);
```
Can we remove println call here? It makes the `komorebi-bar` logs very noisy 😅

davor-skontra replied:
Missed that one, my bad!

LGUG2Z commented on src/uwp_apps.rs:
```diff
@@ -0,0 +1,56 @@
+pub fn get_uwp_icon(process_path: &str) -> Result<RgbaImage, Box<dyn Error>> {
+    let icon_path = get_icon_file_path(process_path)?;
+    println!("Icon path: {}", icon_path);
+    let base64 = read_image_to_base64(&icon_path)?;
```
It looks like either this call or the one beneath it is failing when trying to get the icon for Windows Terminal 🤔

davor-skontra replied:
Hmm, I might have misunderstood what's what. The upstream fix solved the icon issue with Arc browser and I thought that was it. Ill poke around a bit to see if I can get that working.

davor-skontra replied:
OK, I got this working but its still pretty a bit hacky. Previously there was an assumption in this code that the images referenced from the `ApplicationManifest.xml` always have the exact filename as exists in the file system, but that's not correct. Apparently some metadata is appended to those names.

Things like display scaling, localization, contrast etc. 

So `StoreLogo.png` can become  `StoreLogo.scale-125` or even `StoreLogo.scale-125_contrast-black.png` and similar variations.

Currently there is no handling for any of that in this PR, it just gets a list of all files that should match and returns the first one in the list. (See around line 56 in `src/uwp_apps.rs`)

I'll keep working on this but it might be usable as is for the moment if you don't want to wait.

davor-skontra committed:
Remove debug line

davor-skontra committed:
Make scaled icons work

davor-skontra committed:
Clean up debug lines

davor-skontra committed:
Fix a naming annoyance.

davor-skontra committed:
Add icon matching

This lets us pass a parameter describing what kind of icon we want to
get for uwp apps. Currently it only supports working with display scale,
but localization and accessibility optionsi like contrast can be added
later

davor-skontra committed:
Use xml parsing to fetch the 44x44 icon and add app type pattern matching

davor-skontra committed:
Add missing dependencies

davor-skontra committed:
WIP: Add waiting logic while attempting to get process data

davor-skontra committed:
Use sysinfo to get process path

davor-skontra committed:
Add methods for fetching the icon through HWND

davor-skontra committed:
Get real process through enumerating child windows

davor-skontra committed:
Remove unecessary recursion when fetching process id

davor-skontra committed:
Remove the desktop and Universal paths, keep as UWP

The name UWP is wrong for these kinds of apps, afaik windows has both
UWP and WindowsDesktop apps in the same folder, but as upstream refers
to both as UWP I'm keeping that naming here.

davor-skontra committed:
Replace one unwrap with the ? operator

davor-skontra committed:
Remove usings and one println that does not belong

davor-skontra committed:
Remove some unwraps

davor-skontra committed:
Flatten all use statements, remove unecessary ones

davor-skontra committed:
Remove unwrap when parsing regex

davor-skontra committed:
Use file_stem_contains method in one more place to remove an unwarp

davor-skontra commented:
OK, I've cleaned this up to a point I'm happy with. A few unwraps remain. Mainly unwrapping the `to_str()` calls in methods that return result. 

I think it's safe in those but I'm not sure why `to_str()` doesnt just return a `&str` in the first place so I might be wrong.

Also note there are a few new dependencies in there.

- [regex](https://docs.rs/regex/latest/regex/) -> currently used for parsing filenames when figuring out what icon to grab from the existing ones
- [roxmltree](https://github.com/RazrFalcon/roxmltree)-> parsing the ApplicationManifest.xml files to locate the correct icon
- [sysinfo](https://docs.rs/sysinfo/0.3.6/sysinfo/index.html) -> Made it easier to work with processes then using the windows API directly. This one is most avoidable of the three but saved me some time.

I'm open to finding workarounds to avoid new dependencies, but time is hard to come by so hoping we can keep these.

Note that In order to be able to find all icons you need to make use one of the `get_icon_by_hwnd` methods as I didn't find a way to make that work with `process id`.
